runtime/include/chpl-comm-native-atomics.h

```diff
@@ -0,0 +1,168 @@
+DECL_CHPL_COMM_ATOMIC_PUT(real32)
+DECL_CHPL_COMM_ATOMIC_PUT(real64)
+
+//
+// Do a remote atomic load.  The source location is *object on the
+// given node.  The value is returned in *result on the local node.
+// This differs from a regular chpl_comm_get() in that it is coherent
+// with other chpl_comm_atomic_*() operations (but may incur network
+// overhead) even if the source is the calling node.
+//
+#define DECL_CHPL_COMM_ATOMIC_GET(type)                                 \
+  void chpl_comm_atomic_get_ ## type                                    \
+         (void* result, c_nodeid_t node, void* object,                  \
+          int ln, int32_t fn);
+
+DECL_CHPL_COMM_ATOMIC_GET(int32)
+DECL_CHPL_COMM_ATOMIC_GET(int64)
+DECL_CHPL_COMM_ATOMIC_GET(uint32)
+DECL_CHPL_COMM_ATOMIC_GET(uint64)
+DECL_CHPL_COMM_ATOMIC_GET(real32)
+DECL_CHPL_COMM_ATOMIC_GET(real64)
+
+//
+// Do a remote atomic exchange.  The value to be stored is *desired on
+// the local node.  The target location to be stored into is *object
+// on the given node.  The value previously stored there is returned
+// in *result on the local node.
+//
+#define DECL_CHPL_COMM_ATOMIC_XCHG(type)                                \
+  void chpl_comm_atomic_xchg_ ## type                                   \
+         (void* desired, c_nodeid_t node, void* object, void* result,   \
+          int ln, int32_t fn);
+
+DECL_CHPL_COMM_ATOMIC_XCHG(int32)
+DECL_CHPL_COMM_ATOMIC_XCHG(int64)
+DECL_CHPL_COMM_ATOMIC_XCHG(uint32)
+DECL_CHPL_COMM_ATOMIC_XCHG(uint64)
+DECL_CHPL_COMM_ATOMIC_XCHG(real32)
+DECL_CHPL_COMM_ATOMIC_XCHG(real64)
+
+//
+// Do a remote atomic compare and exchange.  The value to be matched
+// is *expected on the local node.  If the match succeeds, the value
+// to be stored is *desired on the local node.  The target location to
+// be stored into is *object on the given node.  Whether the exchange
+// occurred or not is returned in *result on the local node.
+//
+#define DECL_CHPL_COMM_ATOMIC_CMPXCHG(type)                             \
+  void chpl_comm_atomic_cmpxchg_ ## type                                \
+         (void* expected, void* desired, c_nodeid_t node, void* object, \
+          chpl_bool32* result,                                          \
+          int ln, int32_t fn);
+
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(int32)
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(int64)
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(uint32)
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(uint64)
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(real32)
+DECL_CHPL_COMM_ATOMIC_CMPXCHG(real64)
+
+//
+// Do a remote atomic binary operation, non-fetching or fetching.  In
+// either case, the operand is *operand on the local node and the target
+// location is *object on the given node.  For the fetching style, the
+// value the target had prior to the operation is returned in *result on
+// the local node.
+//
+// We support AND, OR, and XOR for integers, and ADD and SUB for ints
+// and reals.  In the future we might like to add other operations, such
```

Comment:
Maybe remove the "in the future note". I don't think it's valuable to have potential future feature requests in comment